### DISCUSSION TITLE
Welcome to jsonapi-resources Discussions!

### DISCUSSION THREAD

**Discussion by lgebhardt** [JSONAPI-Resources/jsonapi-resources#1432] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by ghost** [comment]
Hello! If this kind of question is not appropriate for this discussion area please let me know.

JSON:API allows attributes to be objects, and I would like to be able to do this with jsonapi-resources.

For example, instead of something like this, where the attributes section is completely flat:
```
{
  "data": {
    "id": "1",
    "type": "envelope",
    "attributes": {
      "addressee": "Mr. Smith",
      "to_address": "123 Main St",
      "to_city": "New York City",
      "to_state": "NY"
      "to_zip": "11111"
      "return_address": "456 Fun St",
      "return_city": "Sacramento",
      "return_state": "CA"
      "return_zip": "99999"
    }
  }
}
```

I'd like to be able to do this, where I can group attributes with objects:
```
{
  "data": {
    "id": "1",
    "type": "envelope",
    "attributes": {
      "addressee": "Mr. Smith",
      "to_address": {
        "address": "123 Main St",
        "city": "New York City",
        "state": "NY"
        "zip": "11111"
      },
      "return_address": {
        "address": "456 Fun St",
        "city": "Sacramento",
        "state": "CA"
        "zip": "99999"
      }
    }
  }
}
```

Using this example, I would not want addresses to be separate resources, and don't want a whole bunch of unique attribute names (using prefixes) to differentiate similar but different address parts - I'd rather group them. I can do this with vanilla Rails using ActiveRecord's `composed_of` concept - I'd love advice on how to do this with JR, or get it into the feature roadmap. It seems like the framework could handle this with some modest extra syntax. Thanks for reading!

UPDATE:
I could make this sort of work by introducing an `Address` value object, using ActiveRecord's `composed_of`, and creating a JR `AddressValueFormatter`. The JSON looks exactly as I would want. However other things do not quite fit - such as trying to sort a nested attribute, or to PATCH a single nested attribute (the other fields get set to nil).
